### ISSUE TITLE
Cantera's definition/calculation of rate constants

### ISSUE THREAD

**Issue by ischoegl** [Cantera/cantera#1083]
**Problem description**

In the current code base, third-body concentrations are multiplied with rate constants, see https://github.com/Cantera/cantera/blob/375902b50127d046bc6d1b7e7e6dd86e8fd3d89a/src/kinetics/GasKinetics.cpp#L230-L256 
This approach (lines 238-241) does not follow textbook definitions, or at least not the one used by Kee/Coltrin/Glarborg's text, see Eqs. 9.73 and 9.75 (Chemically Reacting Flow). An implementation consistent with Eq. 9.75 would multiply the effective third-body concentration only for the calculation of rates-of-progress. `Falloff` reactions are a separate case as they use a modified rate constant, which does include third-body concentrations.

**Additional context**

The definition is clearly historic and only affects a few cases where `Kinetics::getFwdRateConstants` and `Kinetics::getRevRateConstants` (or their API's) are used directly in `ThreeBodyReaction`s.

**Comment by bryanwweber** [comment]
Can you clarify why this is an issue, or something that needs to be fixed/changed?

**Comment by ischoegl** [comment]
@bryanwweber ... I believe that Cantera output should follow accepted conventions, which should automatically make it an issue. A 'fix' is straight-forward, with the deprecation of the current behavior - as usual - being the trickiest part. 

For context: while it may not be widely used, I just ran head-on into this during #1081. Despite being aware of this choice, I spent a fair amount of time chasing my tail for some failing unit tests. Rather than accommodating this unusual choice, I believe fixing this 'paper-cut' is ultimately the better option.

**Comment by bryanwweber** [comment]
@ischoegl Thanks for the context! I have a few thoughts here. First, I agree that this feels somewhat unusual and certainly trying to track down a "bug" like that is super frustrating (been there :smile:). That said, I don't think anecdata are enough to warrant a change here; it's hard to be more authoritative than Bob Kee, but I'd want some more demonstration that this approach really is all that unusual.

Second, I suspect the idea with this code is some computational and/or implementation efficiency. For example, if you include the 3rd body concentrations here, then you don't need to do it in the ROP calculation, which becomes a simple multiplication by the reactant concentrations for all reactions. I haven't looked into it, and certainly the code branching has to go _somewhere_, but I wonder if that might be a reason. Another thought is with respect to reverse rate constants, which depend on the forward rate constant - could that be why this is included here?

Third, are we sure that fixing this paper cut isn't going to introduce new paper cuts? For instance (presuming this choice goes back to Dave's original work), changing this would effectively break 20 years worth of Cantera scripts for something that's pretty minor, for anyone who's working directly with the rates. IMO, that's a pretty big resulting paper cut...

**Comment by ischoegl** [comment]
@bryanwweber ... thank you for your thoughts; as always, you have good points. At the same time, I was pretty deep in the weeds of `updateROP`, and there is not really a good reason for speed here - a fix would only shuffle the order a little bit (curiously, the original implementation of `getFwdRateConstants` is only a partial cut&paste of `updateROP`).

Regarding breaking tradition: yes, this code has been around, but it is still not intuitive (or consistent with theory). The mechanics here would involve a relatively straight-forward deprecation. Below is a suggestion:
```
In [1]: import cantera as ct
   ...: gas = ct.Solution("h2o2.yaml")
   ...: gas.TPX = 300, ct.one_atm, 'H2:1, O2:3'
   ...: gas.equilibrate('HP')
   ...: 

In [2]: gas.forward_rate_constants
DeprecationWarning: GasKinetics::getFwdRateConstants: Behavior to change after Cantera 2.6;
results will no longer include third-body concentrations for ThreeBodyReaction objects.
Set 'legacyRateConstants' to 'false' for new behavior.
Out[2]: 
array([1.66615789e+06, 3.30877601e+06, 8.13054905e+09, 2.00000000e+10,
       1.63880887e+10, 2.03131173e+05, 6.47691445e+06, 5.46151483e+07,
       0.00000000e+00, 0.00000000e+00, 2.65917764e+09, 1.97800108e+06,
       1.07713432e+03, 6.84832674e+06, 5.01798085e+07, 3.38149587e+09,
       3.47031524e+10, 7.21665842e+10, 1.54550508e+10, 4.22813794e+09,
       9.91305045e+09, 2.98116515e+07, 5.58903179e+09, 1.63414930e+10,
       1.80587269e+09, 1.50070437e+12, 1.91961755e+08, 2.38275715e+10,
       7.93037160e+10])

In [3]: gas.legacy_rate_constants = False

In [4]: gas.forward_rate_constants
Out[4]: 
array([5.70210643e+07, 2.37587768e+08, 8.13054905e+09, 2.00000000e+10,
       1.63880887e+10, 3.88376134e+09, 1.57529678e+09, 3.35697020e+10,
       1.96912098e+09, 1.53667243e+09, 2.65917764e+09, 4.75175536e+08,
       9.12756539e+08, 4.20938686e+09, 4.96741937e+09, 3.38149587e+09,
       3.47031524e+10, 7.21665842e+10, 1.54550508e+10, 4.22813794e+09,
       9.91305045e+09, 2.98116515e+07, 5.58903179e+09, 1.63414930e+10,
       1.80587269e+09, 1.50070437e+12, 1.91961755e+08, 2.38275715e+10,
       7.93037160e+10])
```
Above, it takes a single line to switch and overall feels like any other deprecation. I included what is above in #1081 (commit 8c92ff25ce21c0bfe658985f7680e0cb89c0cca), although there's a little more refactoring going on. 

There still is more churn than I'd like for the test suite, so a global flag (similar to `ct.make_deprecation_warnings_fatal`) may be the easiest / least intrusive alternative. I'd still like to update this as the current implementation is definitely not what you'd expect from textbook theory. 

**PS:** Another comment here (which I almost forgot) is that it also plays into the ongoing work for the new generic `ReactionRate`  objects. If possible, those should correspond to what you'd find in any standard reference. These objects are accessible from all `Reaction`s in the same fashion and can be evaluated. It would be imho inconsistent if the output of `gas.reaction(ix).rate(gas.T)` were different than `gas.forward_rate_constants[ix]` at a given temperature. Or, to give full example code:
```
In [5]: ix = 0
   ...: gas.reaction(ix)
   ...: 
Out[5]: <ThreeBodyReaction: 2 O + M <=> O2 + M>

In [6]: gas.legacy_rate_constants = True # current implementation

In [7]: gas.reaction(ix).rate(gas.T)
Out[7]: 57021064.2631974

In [8]: gas.forward_rate_constants[ix]
Out[8]: 1666157.893731073

In [9]: gas.legacy_rate_constants = False # proposed implementation

In [10]: gas.reaction(ix).rate(gas.T)
Out[10]: 57021064.2631974

In [11]: gas.forward_rate_constants[ix]
Out[11]: 57021064.2631974
```

**Comment by ischoegl** [comment]
@bryanwweber ... I made the fix a separate PR - see #1084 (where there now is a global switch to make a transition as painless as possible). That way, all proposed changes are clearly visible.

**Comment by decaluwe** [comment]
FYI - I just added a review over on the PR, but wanted to chime in here that I'm generally in favor of the change, *if* we can show that the current approach is authoritatively at odds with how this is described/implemented in the literature (beyond just Kee's [rather authoritative 😂 ] text).
